### PR TITLE
BUG: stats.binomtest: fix precision loss in two-sided test for large n

### PR DESCRIPTION
**Pull request by jirigav**
#### What does this implement/fix?
`stats.binomtest` returns sometimes imprecise results for large `n` (>2^53) due to computation in floats, even when int can be used. 

Fixed by converting to `int64` for the subtraction of `n` and `ix`, then casting back to float before passing to `B.sf()`.

#### Additional information
How to reproduce:
```

from scipy.stats import binom, binomtest
from scipy.stats._binomtest import _binary_search_for_binom_tst
import numpy as np

n = 10000000000000000
k = 1
p = 1e-15

d = binom.pmf(k, n, p)
rerr = 1 + 1e-7

ix = _binary_search_for_binom_tst(
    lambda x1: -binom.pmf(x1, n, p),
    -d * rerr, np.ceil(p * n), n)

y = n - int(ix) + int(d*rerr == binom.pmf(ix, n, p))
pval_correct = binom.cdf(k, n, p) + binom.sf(n - y, n, p)


result = binomtest(k, n, p, 'two-sided')

print(f"binomtest: {result.pvalue}") # 0.002087659889245364
print(f"correct:   {pval_correct}")  # 0.001199049739722154

print(f"n - int(ix) = {n - int(ix)}")
print(f"n - ix = {int(n - ix)}")
```

#### AI Generation Disclosure
Claude AI was used to make the change more canonical.

### CODE REVIEW

File: scipy/stats/tests/test_morestats.py [hunk 1]
```diff
@@ -1466,6 +1466,16 @@ def test_binomtest3(self, xp):
...
 
+    @skip_xp_backends('jax.numpy', reason="'two-sided' alternative needs root finder")
+    def test_binomtest_large_n_precision(self, xp):
+        # Test that two-sided binomtest is precise for very large n,
+        # where float arithmetic on n and ix would lose precision.
+        n = 10000000000000000
+        k = 1
+        p = 1e-15
+        result = stats.binomtest(k, n, p)
+        xp_assert_close(result.pvalue, xp.asarray(0.001199049739722154))
```
Comment (tylerjereddy) [inline scipy/stats/tests/test_morestats.py:1477]: Can you provide us with some information indicating why this new expected value is correct compared to the previous result?

I tried checking `binom.test` in R using https://webr.r-wasm.org/latest/, via `binom.test(x = 1, n = 10000000000000000, p = 1e-15, alternative = "two.sided")`

and it returned:

```
Error in seq.int(from = ceiling(m), to = n) : 
  result would be too long a vector
```

How did you run into this in the first place? Naturally in research or?

**Review by tylerjereddy — COMMENTED** [review]
Can you provide some information about how you ran into this and how you know the tested expected value is correct?

**Comment by mdhaber** [comment]
I'd suggest that we not make a change here. I see no reason to believe that $2^{63}$ is enough if $2^{53}$ isn't.  I know it's fun to squeeze all the precision we can out of 64 bits, but this is precisely the case where arbitrary precision would shine, and we will get that for free when gh-24840 is done.

*Update: I tried it, and basically all that's needed to make `binomtest` work with `mparray` is to adjust some special function delegation. There's a lot more bang for buck to be had there than in value-dependent switching between integral and floating point dtypes. @tylerjereddy the new expected value was correct to more digits because it doesn't lose precision prematurely by representing large integers as floating point values, but the true value is closer to `0.00119904973972215577690442984516564541123152643375535397184315266251348058755446389101935912847742655782388367546610798630241482692199503146318998980975135215170816149946174139800294277225598136088621437762475489744253051304906829655143278483604081964542035251438033327152184156784188623869625682131130562`, although I would only trust the first 250 digits or so.*

File: scipy/stats/tests/test_morestats.py (hunk 1 above)
Comment (jirigav) [reply scipy/stats/tests/test_morestats.py:1477]: I've run into it while fuzzing my implementation of the binomial test in Rust against the scipy implementation. So, it's fair to say that the example is rather artificial. 
I believe that it is fair to assume that the result without the off-by-one error is closer to correct than the original one with the error. But yes, I also tried R and got the same error. 

However, I believe that @mdhaber is right, and it does not make sense to do this fix.

**Comment by jirigav** [comment]
@mdhaber 
I think that you are right and it does not make sense to fix this if arbitrary precision is coming soon.
